File: Sources/XMTP/Codecs/GroupChatMemberAddedCodec.swift
//
//  GroupChatMemberAddedCodec.swift
//
//
//  Created by Pat Nakajima on 6/11/23.
//

import Foundation

public let ContentTypeGroupChatMemberAdded = ContentTypeID(authorityID: "xmtp.org", typeID: "groupChatMemberAdded", versionMajor: 1, versionMinor: 0)

public struct GroupChatMemberAdded: Codable {
	// The address of the member being added
	public var member: String
}

public struct GroupChatMemberAddedCodec: ContentCodec {
	public var contentType = ContentTypeGroupChatMemberAdded

	public func encode(content: GroupChatMemberAdded) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeGroupChatMemberAdded
		encodedContent.content = try JSONEncoder().encode(content)

		return encodedContent
	}

	public func decode(content: EncodedContent) throws -> GroupChatMemberAdded {
		let memberAdded = try JSONDecoder().decode(GroupChatMemberAdded.self, from: content.content)
		return memberAdded
	}
}

File: library/src/main/java/org/xmtp/android/library/codecs/GroupChatMemberAddedCodec.kt
package uniffi.xmtp_dh.org.xmtp.android.library.codecs

import com.google.gson.GsonBuilder
import com.google.protobuf.kotlin.toByteStringUtf8
import org.xmtp.android.library.codecs.ContentCodec
import org.xmtp.android.library.codecs.ContentTypeId
import org.xmtp.android.library.codecs.ContentTypeIdBuilder
import org.xmtp.android.library.codecs.EncodedContent

val ContentTypeGroupChatMemberAdded = ContentTypeIdBuilder.builderFromAuthorityId(
    "xmtp.org",
    "groupChatMemberAdded",
    versionMajor = 1,
    versionMinor = 0
)

// The address of the member being added
data class GroupChatMemberAdded(var member: String)

data class GroupChatMemberAddedCodec(override var contentType: ContentTypeId = ContentTypeGroupChatMemberAdded) :
    ContentCodec<GroupChatMemberAdded> {

    override fun encode(content: GroupChatMemberAdded): EncodedContent {
        val gson = GsonBuilder().create()
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeGroupChatMemberAdded
            it.content = gson.toJson(content).toByteStringUtf8()
        }.build()
    }

    override fun decode(content: EncodedContent): GroupChatMemberAdded {
        val gson = GsonBuilder().create()
        return gson.fromJson(content.content.toStringUtf8(), GroupChatMemberAdded::class.java)
    }
}
